java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaBucketListingIterator.java

```diff
@@ -0,0 +1,332 @@
+
+    /**
+     * The last marker we used to request against the Manta API.
+     */
+    private volatile String lastMarker;
+
+    /**
+     * The current {@link BufferedReader} instance that wraps the HTTP response
+     * {@link java.io.InputStream} from our most recent request to the API.
+     */
+    private volatile BufferedReader br;
+
+    /**
+     * The most recent response object from the page of data that we are currently
+     * parsing.
+     */
+    private volatile CloseableHttpResponse currentResponse;
+
+    /**
+     * Create a new instance of a bucket list iterator.
+     *
+     * @param path path to bucket in which we will iterate through its contents
+     * @param httpHelper HTTP request helper class
+     * @param isSorted true for lexicographical ordering of buckets
+     * @param prefix filter for listing buckets
+     */
+    public MantaBucketListingIterator(final String path,
+                                      final HttpHelper httpHelper,
+                                      final boolean isSorted,
+                                      final String prefix) {
+        Validate.notBlank(path, "Path must not be blank");
+        Validate.notNull(httpHelper, "HTTP help must not be null");
+
+        this.path = path;
+        this.httpHelper = httpHelper;
+
+        if (MantaUtils.parseBooleanOrNull(isSorted)) {
+            throw new IllegalArgumentException("Invalid query parameter supplied"
+                    + "for sorted buckets listing");
+        }
+
+        // Conditional check for prefix if needed
+
+        this.isSorted = isSorted;
+        this.prefix = prefix;
+    }
+
+    /**
+     * Create a new instance of a bucket list iterator.
+     *
+     * @param url ignored
+     * @param path path to bucket in which we will iterate through its contents
+     * @param httpHelper HTTP request helper class
+     */
+    public MantaBucketListingIterator(final String url,
+                                         final String path,
+                                         final HttpHelper httpHelper) {
+        this(path, httpHelper, false, DEFAULT_PREFIX);
+    }
+
+    /**
+     * Create a new instance of a bucket list iterator.
+     *
+     * @param path path to bucket in which we will iterate through its contents
+     * @param httpHelper HTTP request helper class
+     */
+    public MantaBucketListingIterator(final String path,
+                                         final HttpHelper httpHelper) {
+        this(path, httpHelper, false, DEFAULT_PREFIX);
+    }
+
+
+    /**
+     * Chooses the next reader by opening a HTTP connection to get the next
+     * page of input from the Manta API. If there isn't another page of data
+     * available, we mark ourselves as finished.
+     *
+     * @throws IOException thrown when we can't successfully open an HTTP connection
+     */
+    private synchronized void selectReader() throws IOException {
+        if (lastMarker == null) {
```

Comment:
How would this method incorporate the ```marker``` option in the request? Also will it be using the ```Next-Marker``` field in the response to gauge if there are more objects present?
```
$ manta "/$MANTA_USER/buckets?limit=100&marker=z-bucket" | json -ag name
z-bucket-01
z-bucket-02
z-bucket-03
z-bucket-04
z-bucket-05
$ manta "/$MANTA_USER/buckets?limit=1&marker=z-bucket-02" -i | egrep '^({|Next-Marker)'
Next-Marker: z-bucket-03
{"name":"z-bucket-03","type":"bucket","mtime":"2019-07-17T20:21:59.637Z"}
$ manta "/$MANTA_USER/buckets?limit=1&marker=z-bucket-03" -i | egrep '^({|Next-Marker)'
Next-Marker: z-bucket-04
{"name":"z-bucket-04","type":"bucket","mtime":"2019-07-17T20:22:03.033Z"}
$ manta "/$MANTA_USER/buckets?limit=1&marker=z-bucket-04" -i | egrep '^({|Next-Marker)'
{"name":"z-bucket-05","type":"bucket","mtime":"2019-07-17T20:22:06.192Z"}
```